packages/shared/src/lib/textarea.ts

```diff
@@ -0,0 +1,154 @@
+import { isNullOrUndefined, nextTick } from './func';
+
+export const isFalsyOrSpace = (value: string): boolean =>
+  !value || value === ' ';
+
+const getSelectedString = (textarea: HTMLTextAreaElement) => {
+  const [start, end] = [textarea.selectionStart, textarea.selectionEnd];
+  const last = start === end ? start + 1 : end;
+  const text = textarea.value.substring(start, last);
+  const before = textarea.value[start - 1];
+
+  return [text, before];
+};
+
+const splitString = (value: string, [start, end]: number[]) => {
+  const left = value.substring(0, start);
+  const right = value.substring(end);
+
+  return [left, right];
+};
+
+const concatReplacement = (
+  textarea: HTMLTextAreaElement,
+  selected: number[],
+  replace: string,
+) => {
+  const [left, right] = splitString(textarea.value, selected);
+
+  return left + replace + right;
+};
+
+/* eslint-disable no-param-reassign */
+const focusInput = async (
+  textarea: HTMLTextAreaElement,
+  [start, end]: number[],
+) => {
+  textarea.focus();
+  await nextTick(); // for the selection to work, we need to wait for the input to get updated
+  textarea.selectionStart = start;
+  textarea.selectionEnd = end;
+};
+
+export const getCloseWord = (
+  textarea: HTMLTextAreaElement,
+  selected: number[],
+): [string, number] => {
+  const [, end] = selected;
+
+  let lastIndex = 0;
+  const words = textarea.value.split(' ');
+  const closeWord = words.find((word) => {
+    const current = lastIndex + word.length + 1;
+
+    if (current > end) return true;
+
+    lastIndex = current;
+
+    return false;
+  });
+
+  return [closeWord?.trimEnd() ?? '', lastIndex];
+};
+
+/**
+ * Enum to describe the current cursor's state.
+ * The character `|` denotes the cursor
+ * @enum CursorType
+ * Isolated: when the cursor is in-between spaces (ex. " | ").
+ * Adjacent: when the cursor is beside any character (ex. "Test |word" or "Test| word").
+ * Highlighted: when the cursor is highlighting any string (ex. "|Tes|t" or "|Test|").
+ */
+export enum CursorType {
+  Isolated = 'isolated',
+  Adjacent = 'adjacent',
+  Highlighted = 'highlighted',
+}
+
+const getCursorType = (textarea: HTMLTextAreaElement) => {
+  const [start, end] = [textarea.selectionStart, textarea.selectionEnd];
+  const [highlighted, before] = getSelectedString(textarea);
+
+  if (isFalsyOrSpace(highlighted) && isFalsyOrSpace(before)) {
+    return CursorType.Isolated;
+  }
+
+  return start !== end ? CursorType.Highlighted : CursorType.Adjacent;
+};
+
+export interface GetReplacement {
+  addSpaceBeforeHighlighted?: boolean;
+  offset?: number[];
+  replacement: string;
+}
+
+interface GetReplacementOptionalProps {
+  word?: string;
+  trailingChar?: string;
+}
+
+export type GetReplacementFn = (
+  type: CursorType,
+  word?: GetReplacementOptionalProps,
+) => GetReplacement;
+
+export const replaceWord = async (
```

Comment:
This is where it all happens. There should not be too much change here in the future as its purpose is simple, to replace the word based on the parameters. The optional parameter is how to set the cursor position if the default behavior is not what you would expect.

Reply:
Agree with that. Let me see how we can break this down into pieces.